src/Sarif/ResultsCachingLogger.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Sarif
+{
+    public class ResultsCachingLogger : IAnalysisLogger
+    {
+        private bool cacheLoggingData;
+        private string currentFileHash;
+        
+        public Dictionary<string, List<Tuple<ReportingDescriptor, Result>>> HashToResultsMap { get; private set; }
+
+        public ResultsCachingLogger(bool verbose)
+        {
+            Verbose = verbose;
+        }
+
+        public void AnalysisStarted()
+        {
+            HashToResultsMap = new Dictionary<string, List<Tuple<ReportingDescriptor, Result>>>();
+        }
+
+        public bool Verbose { get; private set; }
+
+        public void AnalysisStopped(RuntimeConditions runtimeConditions)
+        {
+        }
+
+        public void AnalyzingTarget(IAnalysisContext context)
+        {
+            this.currentFileHash = context.Hashes.Sha256;
+
+            if (HashToResultsMap.ContainsKey(currentFileHash))
+            {
+                // We already have results cached for this file. This means that we shouldn't collect
+                // any data in calls to Log. These calls are the driver framework 'replaying' the
+                // cached results via the aggregated logger instance
+                cacheLoggingData = false;
+            }
+            else
+            {
+                cacheLoggingData = true;
+                HashToResultsMap[currentFileHash] = new List<Tuple<ReportingDescriptor, Result>>();
+            }
+        }
+
+        public void Log(ReportingDescriptor rule, Result result)
+        {
+            if (!cacheLoggingData) { return; }
+
+            switch (result.Level)
+            {
+                // These result types are optionally emitted.
```

Comment:
>These [](start = 19, length = 5)

It's a pity you have to remember that to write this code correctly. It duplicates logic elsewhere. If we introduced more verbosity levels ("only errors", for example), we'd have to remember to change this code that decides whether to cache the result.

Is there any harm in just always caching it, regardless of verbosity level?

Reply:
Or perhaps introduce introduce a utility method `Utilities.ResultShouldBeLogged(level, verbosity)` and call it in both places.

---
In reply to: [303215190](https://github.com/microsoft/sarif-sdk/pull/1574#discussion_r303215190) [](ancestors = 303215190)